File: polyglot/ruby/git_scanner.rb
# frozen_string_literal: true

require 'bundler'
require 'json'
require 'open3'
require 'pathname'
require 'set'
require 'time'
require 'uri'

class Pipewatch::GitScanner
  # OWASP CI/CD Top 10 patterns to scan for in commits and code
  class << self
    attr_reader :owasp_patterns, :dangerous_keywords, :common_dep_patterns
    
    def initialize(repo_path: '.', options: {})
      @repo_path = Pathname.new(repo_path)
      @options = {
        verbose: false,
        include_submodules: true,
        check_commits: true,
        check_code: true,
        **options
      }.compact
      
      # Initialize pattern sets
      @owasp_patterns = build_owasp_patterns
      @dangerous_keywords = Set.new([
        'eval', 'system', 'exec', `cat`, 'shell_exec', 
        'popen', 'spawn', 'backtick', 'cmd.exe', 
        'powershell', 'wget', 'curl', 'ncat', 'netcat'
      ].flatten)
      @common_dep_patterns = Set.new([
        'gem install', 'npm install', 'yarn add', 
        'pip install', 'cargo add', 'bundle install',
        'composer require', 'apt-get install'
      ])
    end
    
    private

    def build_owasp_patterns
      patterns = []
      
      # A01:2021 - Broken Access Control (path traversal, etc)
      patterns << /A01.*?(?:traversal|injection|bypass)/i
      
      # A02:2021 - Cryptographic Failures
      patterns << /A02.*?(?:weak|default|hardcoded|secret|key|token)/i
      
      # A03:2021 - Design Flaws
      patterns << /A03.*?(?:race|concurrency|buffer|overflow)/i
      
      # A04:2021 - External Dependencies
      patterns << /A04.*?(?:supply|chain|vendor|third-party)/i
      
      # A05:2021 - Identification & Authentication Failures
      patterns << /A05.*?(?:session|cookie|auth|login|password)/i
      
      # A06:2021 - Security Misconfiguration
      patterns << /A06.*?(?:debug|dev|verbose|log|trace)/i
      
      # A07:2021 - XSS (Cross-Site Scripting)
      patterns << /A07.*?(?:xss|script|javascript|on\w+|innerHTML)/i
      
      # A08:2021 - Insufficient Logging & Monitoring
      patterns << /A08.*?(?:log|monitor|audit|track)/i
      
      # A09:2021 - Security Logging and Monitoring Failures
      patterns << /A09.*?(?:silent|quiet|suppress|capture)/i
      
      # A10:2021 - Server-Side Request Forgery (SSRF)
      patterns << /A10.*?(?:ssrf|fetch|request|curl|wget)/i
      
      Set.new(patterns.compact.uniq)
    end
    
    def run_scan
      results = {
        repo_path: @repo_path.to_s,
        scan_time: Time.now.iso8601,
        version: Pipewatch::GitScanner::VERSION,
        summary: {},
        details: {}
      }
      
      if !@repo_path.exist?
        results[:errors] = ["Repository not found: #{@repo_path}"]
        return results
      end
      
      # Check if it's a git repository
      unless @repo_path.join('.git').exist? || 
             @repo_path.join('HEAD').readable?
        results[:warnings] << "May not be a valid Git repository"
      else
        run_git_metadata_scan(results)
        run_commit_analysis(results) if @options[:check_commits]
        run_code_analysis(results) if @options[:check_code]
      end
      
      # Generate summary
      results[:summary] = build_summary(results)
      
      results
    rescue StandardError => e
      results[:errors] << "Scan failed: #{e.message}"
      results[:error_backtrace] = e.backtrace.first(10).join("\n")
      results
    end
    
    def run_git_metadata_scan(results)
      begin
        # Get basic repo info
        output, _status, _err = Open3.capture3('git', 'rev-parse', '--show-toplevel')
        
        if $?.success? && output.strip != @repo_path.to_s
          results[:metadata] = {
            root: output.strip,
            branch: `git rev-parse --abbrev-ref HEAD`.strip,
            commit: `git rev-parse --short HEAD`.strip,
            author: `git log -1 --format='%an'`.strip,
            date: `git log -1 --format='%ad' --date=iso8601`.strip,
            total_commits: `git rev-list --count HEAD`.to_i
          }
        end
        
        # Check for submodules
        results[:metadata][:submodules] = [] if @options[:include_submodules]
        output, _status, _err = Open3.capture3('git', 'config', '--list', '-z')
        
        if $?.success? && output.include?('submodule')
          results[:metadata][:has_submodules] = true
        end
        
      rescue StandardError => e
        results[:warnings] << "Git metadata error: #{e.message}"
      end
      
      results
    end
    
    def run_commit_analysis(results)
      begin
        # Get recent commits for analysis
        commit_limit = @options.fetch(:commit_limit, 100)
        
        output, _status, _err = Open3.capture3(
          'git', 'log', "--format=%H%n%B%n%an <%ae> %ad", 
          '-n', "#{commit_limit}^..HEAD"
        )
        
        if $?.success? && !output.empty?
          commits = output.split("\n").map.with_index do |line, idx|
            next unless line.strip.length > 0
            
            # Parse commit info (simplified format)
            parts = line.split("\n")
            hash = parts[0].strip if parts[0] && !parts[0].include?('\n')
            
            {
              index: idx,
              hash: hash&.strip,
              message: parts[1..-2]&.join("\n").strip,
              author: parts[-1]&.split(' ').first,
              email: parts[-1]&.include?('<') ? 
                parts[-1].match(/<([^>]+)>/) ? Regexp.last_match(1).strip : nil : nil,
              date: Time.now.iso8601 # Simplified
            }
          end.compact.reject { |c| c[:hash].nil? }
        
        results[:commits] = commits
        
        # Analyze each commit for OWASP patterns
        if @options[:check_commits] && !@owasp_patterns.empty?
          analyze_commits_for_owasp(commits, results)
        end
        
      rescue StandardError => e
        results[:warnings] << "Commit analysis error: #{e.message}"
      end
      
      results
    end
    
    def run_code_analysis(results)
      begin
        # Scan codebase for dangerous patterns and dependencies
        files_scanned = 0
        lines_scanned = 0
        
        # Find relevant source files
        extensions = ['rb', 'js', 'ts', 'py', 'java', 'c', 'h', 'php', 'go']
        
        output, _status, _err = Open3.capture3(
          'git', 'ls-files', '-z'
        )
        
        if $?.success? && !output.empty?
          files = output.split("\0").reject { |f| f.include?('node_modules') || 
                                                    f.include?('vendor') }
          
          # Limit for performance
            max_files = @options.fetch(:max_files, 500)
            files = files.first(max_files) if files.length > max_files
            
        results[:metadata][:files_scanned] = files.length
        
        # Analyze each file
        files.each do |file|
          next unless file.include?('.') && !file.start_with?('.')
          
          full_path = @repo_path.join(file)
          next unless full_path.exist?
          
            begin
              content = File.read(full_path, encoding: 'UTF-8')
              lines_scanned += 1
              
              # Check for OWASP patterns in code
              if @options[:check_code] && !@owasp_patterns.empty?
                analyze_file_for_owasp(content, file, results)
              end
              
              # Extract dependencies from comments and imports
              extract_dependencies(content, file, results)
              
            rescue StandardError => e
              next unless e.is_a?(ArgumentError) || e.is_a?(EncodingError)
              results[:warnings] << "File error: #{file} - #{e.message}"
            end
            
          files_scanned += 1
        end
        
      rescue StandardError => e
        results[:warnings] << "Code analysis error: #{e.message}"
      end
      
      results
    end
    
    def analyze_commits_for_owasp(commits, results)
      commit_issues = []
      
      commits.each do |commit|
        next unless commit[:message] && !commit[:message].empty?
        
        # Check message for OWASP patterns
        @owasp_patterns.each do |pattern|
          if pattern.match?(commit[:message])
            issue = {
              type: 'OWASP Pattern',
              file: nil,
              line: nil,
              commit_hash: commit[:hash],
              commit_message: commit[:message].strip(200),
              pattern: pattern.source,
              severity: 'medium'
            }
            
            # Determine severity based on pattern type
            if /A01|A04/i.match?(pattern) || /traversal|injection/i.match?(commit[:message])
              issue[:severity] = 'high'
            elsif /A02|A05|A07/i.match?(pattern)
              issue[:severity] = 'medium'
            end
            
            commit_issues << issue
          end
        end
        
        # Check for dangerous keywords in message
        if @dangerous_keywords.any? { |kw| commit[:message].include?(kw) }
          issue = {
            type: 'Dangerous Keyword',
            file: nil,
            line: nil,
            commit_hash: commit[:hash],
            commit_message: commit[:message].strip(200),
            keyword: @dangerous_keywords.first,
            severity: 'low'
          }
          
          commit_issues << issue if !commit_issues.any? { |i| i[:keyword] == issue[:keyword] }
        end
      end
      
      results[:details][:owasp_commit_findings] = commit_issues unless commit_issues.empty?
    end
    
    def analyze_file_for_owasp(content, filename, results)
      return if content.length > 50_000 # Performance limit
      
      file_issues = []
      
      @owasp_patterns.each do |pattern|
        matches = pattern.scan(content).flatten.compact.uniq
        next unless matches.any?
        
        issue = {
          type: 'OWASP Pattern',
          file: filename,
          line: nil, # Would need more parsing for exact lines
          commit_hash: nil,
          commit_message: nil,
          pattern: pattern.source,
          severity: 'medium'
        }
        
        issue[:severity] = 'high' if /A01|A04|A07/i.match?(pattern) || 
                                    matches.any? { |m| m.include?('eval') || m.include?('system') }
        
        file_issues << issue
      end
      
      # Check for hardcoded secrets (simplified)
      secret_patterns = [
        /(?i)(password|secret|api_key|token).*[:=]\s*['"][^'"]{4,}['"]/i,
        /(?i)(private.*key|ssl.*cert|pem).*\s*['"][^'"]{8,}['"]/i
      ]
      
      secret_patterns.each do |pattern|
        if pattern.match?(content)
          issue = {
            type: 'Potential Hardcoded Secret',
            file: filename,
            line: nil,
            commit_hash: nil,
            commit_message: nil,
            pattern: pattern.source,
            severity: 'high'
          }
          
          file_issues << issue
        end
      end
      
      results[:details][:owasp_file_findings] ||= []
      results[:details][:owasp_file_findings].concat(file_issues) unless file_issues.empty?
    end
    
    def extract_dependencies(content, filename, results)
      dep_patterns = [
        /gem\s+['"]([^'"]+)['"]/i,           # Ruby gems
        /require\s*['"]([^'"]+)['"]/i,       # Ruby requires
        /import\s+['"]([^'"]+)['"]/i,        # Python/JS imports
        /from\s+['"]([^'"]+)['"]/i,          # Python from
        /use\s+['"]([^'"]+)['"]/i            # JS use statements
      ]
      
      found_deps = []
      
      dep_patterns.each do |pattern|
        matches = pattern.scan(content).flatten.compact.uniq
        next unless matches.any?
        
        matches.each do |match|
          next if match.length < 3 || match.include?('..')
          
          # Skip common stdlib/stdlib modules
          skip_prefixes = ['bundler', 'rails', 'action', 'active', 
                          'json', 'yaml', 'date', 'time', 'uri']
          
          next if skip_prefixes.any? { |p| match.start_with?(p) }
          
          found_deps << match.strip
        end
      end
      
      results[:details][:dependencies] ||= []
      results[:details][:dependencies].concat(found_deps) unless found_deps.empty?
    end
    
    def build_summary(results)
      summary = {
        total_files: results[:metadata]&.fetch(:files_scanned, 0),
        total_commits: results[:metadata]&.fetch(:total_commits, 0),
        owasp_findings: 0,
        dependency_findings: 0,
        high_severity: 0,
        medium_severity: 0,
        low_severity: 0
      }
      
      # Count findings by severity
      all_findings = []
      
      results[:details].each do |key, value|
        next unless key.start_with?('owasp') || key == 'dependencies'
        
        if value.is_a?(Array)
          all_findings.concat(value)
        elsif value.respond_to?(:values_at)
          all_findings.concat(value.values_at('file', 'commit_hash').compact.flatten.compact.uniq)
        end
      end
      
      # Count by severity
      all_findings.each do |finding|
        next unless finding.is_a?(Hash) && finding[:severity]
        
        summary[finding[:severity].to_s + '_severity'] += 1 if ['high', 'medium', 'low'].include?(finding[:severity])
        summary[:owasp_findings] += 1
      end
      
      # Count dependencies (unique)
      deps = results[:details]&.fetch(:dependencies, []) || []
      unique_deps = deps.uniq.length
      summary[:dependency_findings] = unique_deps
      
      summary
    end
    
    def self::VERSION
      '0.1.0'
    end
  end

# Demo/Entry Point
if __FILE__ == $PROGRAM_NAME
  require 'colorize'
  
  puts "Pipewatch Pro - Git Scanner".magenta.bold
  puts "=" * 50
  
  # Default: scan current directory if no argument provided
  repo_path = ARGV.first || '.'
  
  scanner = Pipewatch::GitScanner.new(repo_path: repo_path, options: {
    verbose: true,
    commit_limit: 50
  })
  
  puts "Scanning: #{repo_path}".cyan
  
  results = scanner.run_scan
  
  # Output summary
  if results[:errors]
    puts "\nErrors:".red.bold
    results[:errors].each { |e| puts "  - #{e}" }
  end
  
  if results[:warnings]
    puts "\nWarnings:".yellow.bold
    results[:warnings].first(5).each { |w| puts "  - #{w}" }
  end
  
  summary = results[:summary]
  
  puts "\nScan Summary".green.bold
  puts "-" * 30
  
  puts "Files scanned:     #{summary[:total_files]}"
  puts "Commits analyzed:  #{summary[:total_commits]}"
  puts "OWASP findings:    #{summary[:owasp_findings]}"
  puts "Dependencies found: #{summary[:dependency_findings]}"
  puts "High severity:     #{summary[:high_severity] || 0}"
  puts "Medium severity:   #{summary[:medium_severity] || 0}"
  puts "Low severity:      #{summary[:low_severity] || 0}"

File: polyglot/rust/git_scanner.rs
use git2::{Repository, RemoteCallbacks, FetchOptions, ObjectType};
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};
use std::time::Duration;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum GitScannerError {
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
    #[error("Git error: {0}")]
    Git(String),
    #[error("Repository not found: {0}")]
    NotFound(PathBuf),
    #[error("Clone failed for: {0}, reason: {1}")]
    CloneFailed(PathBuf, String),
}

pub type Result<T> = std::result::Result<T, GitScannerError>;

#[derive(Debug, Clone)]
pub struct CommitInfo {
    pub id: String,
    pub author_name: String,
    pub email: String,
    pub timestamp: chrono::DateTime<chrono::Utc>,
    pub message: String,
}

#[derive(Debug, Default)]
pub struct ScanReport {
    pub repo_url: Option<String>,
    pub commit_count: u64,
    pub total_authors: usize,
    pub suspicious_patterns: Vec<SuspiciousPattern>,
    pub owasp_findings: Vec<OWASPFinding>,
    pub last_commit_date: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug)]
pub struct SuspiciousPattern {
    pub pattern_type: String,
    pub match_count: u32,
    pub examples: Vec<String>,
}

#[derive(Debug)]
pub struct OWASPFinding {
    pub category: String,
    pub severity: SeverityLevel,
    pub description: String,
    pub commit_id: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SeverityLevel {
    Low,
    Medium,
    High,
    Critical,
}

impl From<SeverityLevel> for &'static str {
    fn from(level: SeverityLevel) -> Self {
        match level {
            SeverityLevel::Low => "LOW",
            SeverityLevel::Medium => "MEDIUM",
            SeverityLevel::High => "HIGH",
            SeverityLevel::Critical => "CRITICAL",
        }
    }
}

pub struct GitScanner {
    repo: Option<Repository>,
    base_path: PathBuf,
}

impl GitScanner {
    pub fn new(repo_url: &str) -> Result<Self> {
        let repo = Repository::init(&PathBuf::from("/tmp/pipewatch-scan"))?;
        Ok(Self {
            repo: Some(repo),
            base_path: PathBuf::from("/tmp/pipewatch-scan"),
        })
    }

    pub fn open_existing(path: &str) -> Result<Self> {
        let repo = Repository::open(path)?;
        Ok(Self {
            repo: Some(repo),
            base_path: PathBuf::from(path),
        })
    }

    pub fn clone_repository(&mut self, url: &str) -> Result<()> {
        if self.repo.is_some() {
            let _ = self.repo.take();
        }
        
        let repo = Repository::init(&self.base_path)?;
        let mut fetch_opts = FetchOptions::new();
        fetch_opts.remote_callbacks(RemoteCallbacks::new().progress(|_, _, _, remaining| {
            if let Ok(remaining) = remaining {
                eprintln!("Cloning: {} / {}", url, remaining);
            }
            Ok(())
        }));

        repo.fetch(url, Some(&mut fetch_opts), None)?;
        
        // Get the remote URL for reporting
        let url = repo.remote_url()?;
        self.repo = Some(repo);
        self.base_path = repo.workdir().map(|p| p.to_path_buf()).unwrap_or_default();
        
        Ok(())
    }

    pub fn get_commit_info(&self, commit_id: &str) -> Result<CommitInfo> {
        let repo = self.repo.as_ref().ok_or(GitScannerError::Git("No repository open".to_string()))?;
        
        let obj = repo.rev_parse(commit_id)?;
        let (author_name, email, timestamp) = match obj.tree() {
            Ok(tree) => {
                let commit = repo.find_commit(obj.id())?;
                let author = commit.author();
                
                let name = author.name().unwrap_or("Unknown");
                let email = author.email().unwrap_or("");
                let timestamp = chrono::DateTime::from_timestamp(
                    author.timestamp(),
                    0,
                ).ok_or_else(|| GitScannerError::Git(format!("Invalid timestamp: {}", commit_id)))?;
                
                (name.to_string(), email.to_string(), timestamp)
            }
            Err(_) => {
                let commit = repo.find_commit(obj.id())?;
                let author = commit.author();
                (author.name().unwrap_or("Unknown").to_string(), 
                 author.email().unwrap_or("").to_string(),
                 chrono::DateTime::from_timestamp(author.timestamp(), 0).ok_or_else(|| GitScannerError::Git(format!("Invalid timestamp: {}", commit_id))?)
            }
        };

        let message = obj.tree().map(|t| {
            t.iter()
                .filter_map(|e| e.value().path())
                .collect::<Vec<_>>()
                .join(", ")
        }).unwrap_or_else(|| "unknown".to_string());

        Ok(CommitInfo {
            id: commit_id.to_string(),
            author_name,
            email,
            timestamp,
            message,
        })
    }

    pub fn scan_commit_message(&self, message: &str) -> Vec<SuspiciousPattern> {
        let mut patterns = Vec::new();
        
        // OWASP Top 10 related keywords
        let owasp_keywords = [
            ("XSS", "Cross-Site Scripting"),
            ("CSRF", "Cross-Site Request Forgery"),
            ("SQLi", "SQL Injection"),
            ("SSRF", "Server-Side Request Forgery"),
            ("XXE", "XML External Entity"),
            ("IDOR", "Insecure Direct Object Reference"),
            ("RCE", "Remote Code Execution"),
            ("DoS", "Denial of Service"),
        ];

        for (keyword, description) in owasp_keywords {
            if message.contains(keyword) {
                patterns.push(SuspiciousPattern {
                    pattern_type: format!("OWASP_{}", keyword),
                    match_count: 1,
                    examples: vec![message.to_string()],
                });
            }
        }

        // Check for common secrets/keys in commit messages
        let secret_patterns = [
            ("API_KEY", "Potential API key exposure"),
            ("SECRET_", "Potential secret variable"),
            ("PASSWORD=", "Potential password assignment"),
            ("TOKEN=", "Potential token assignment"),
        ];

        for (pattern, description) in secret_patterns {
            if message.contains(pattern) {
                patterns.push(SuspiciousPattern {
                    pattern_type: format!("SECRET_{}", pattern),
                    match_count: 1,
                    examples: vec![message.to_string()],
                });
            }
        }

        // Check for common file paths that might indicate issues
        let path_patterns = [
            ("config/", "Configuration directory"),
            (".env", "Environment variables file"),
            ("secrets/", "Secrets directory"),
            ("credentials/", "Credentials directory"),
        ];

        for (pattern, description) in path_patterns {
            if message.contains(pattern) {
                patterns.push(SuspiciousPattern {
                    pattern_type: format!("PATH_{}", pattern),
                    match_count: 1,
                    examples: vec![message.to_string()],
                });
            }
        }

        // Deduplicate and limit results
        let mut seen = HashSet::new();
        patterns.retain(|p| {
            if !seen.insert(&p.pattern_type) {
                p.match_count += 1;
                false
            } else {
                true
            }
        });

        // Limit examples to prevent huge output
        for pattern in &mut patterns {
            pattern.examples.truncate(3);
        }

        patterns
    }

    pub fn scan_commit_message_paths(&self, message: &str) -> Vec<SuspiciousPattern> {
        let mut patterns = Vec::new();
        
        // Check for common vulnerable file paths in commit messages
        let path_keywords = [
            ("node_modules/", "NPM dependency directory"),
            ("vendor/", "Vendor dependencies"),
            ("gems/", "Ruby gems"),
            ("requirements.txt", "Python requirements"),
            ("package.json", "Node.js package manifest"),
            ("Cargo.toml", "Rust dependencies"),
        ];

        for (path, description) in path_keywords {
            if message.contains(path) {
                patterns.push(SuspiciousPattern {
                    pattern_type: format!("PATH_{}", path),
                    match_count: 1,
                    examples: vec![message.to_string()],
                });
            }
        }

        // Check for common vulnerable functions in commit messages
        let func_keywords = [
            ("eval(", "Potential eval usage"),
            ("exec(", "Potential exec usage"),
            ("system(", "Potential system call"),
            ("shell_exec", "Shell execution"),
            ("cmd.exe", "Windows command execution"),
        ];

        for (func, description) in func_keywords {
            if message.contains(func) {
                patterns.push(SuspiciousPattern {
                    pattern_type: format!("FUNC_{}", func),
                    match_count: 1,
                    examples: vec![message.to_string()],
                });
            }
        }

        // Deduplicate
        let mut seen = HashSet::new();
        patterns.retain(|p| !seen.insert(&p.pattern_type));

        for pattern in &mut patterns {
            pattern.examples.truncate(3);
        }

        patterns
    }

    pub fn get_commit_stats(&self) -> Result<ScanReport> {
        let repo = self.repo.as_ref().ok_or(GitScannerError::Git("No repository open".to_string()))?;
        
        // Get commit count
        let revwalk = repo.rev_walk(revit2::RevWalk::DEFAULT)?;
        let commit_count: u64 = revwalk.count();
        
        // Get unique authors
        let mut author_set = HashSet::new();
        for _ in 0..commit_count.min(100) { // Limit to first 100 commits for performance
            if let Some(commit_id) = revwalk.next() {
                let commit = repo.find_commit(commit_id)?;
                let author_name = commit.author().name();
                author_set.insert(author_name);
            }
        }

        // Get last commit date
        let last_commit_date = if commit_count > 0 {
            revwalk.next()
                .and_then(|id| repo.find_commit(id).ok())
                .map(|c| c.author().timestamp())
                .and_then(|ts| chrono::DateTime::from_timestamp(ts, 0).ok())
        } else {
            None
        };

        // Scan commit messages for patterns
        let mut suspicious_patterns = Vec::new();
        let mut owasp_findings = Vec::new();

        for _ in 0..commit_count.min(50) {
            if let Some(commit_id) = revwalk.next() {
                if let Ok(commit) = repo.find_commit(commit_id) {
                    let message = commit.message().unwrap_or("no message");
                    
                    // Scan for OWASP patterns
                    let owasp_patterns = self.scan_commit_message(message);
                    for pattern in &owasp_patterns {
                        suspicious_patterns.push(pattern.clone());
                        
                        if pattern.pattern_type.starts_with("OWASP_") {
                            owasp_findings.push(OWASPFinding {
                                category: "OWASP Top 10".to_string(),
                                severity: SeverityLevel::Medium,
                                description: format!("Found OWASP pattern in commit {}", commit_id),
                                commit_id: Some(commit_id.to_string()),
                            });
                        }
                    }

                    // Scan for path patterns
                    let path_patterns = self.scan_commit_message_paths(message);
                    suspicious_patterns.extend(path_patterns);
                }
            }
        }

        Ok(ScanReport {
            repo_url: repo.remote_url().ok(),
            commit_count,
            total_authors: author_set.len(),
            suspicious_patterns,
            owasp_findings,
            last_commit_date,
        })
    }

    pub fn get_file_tree(&self) -> Result<Vec<String>> {
        let repo = self.repo.as_ref().ok_or(GitScannerError::Git("No repository open".to_string()))?;
        
        // Get the latest commit's tree
        let obj = repo.head()?.peel(ObjectType::Tree)?;
        let tree = obj.tree()?;
        
        let mut files: Vec<_> = tree.iter()
            .map(|e| e.value().path())
            .collect();
        
        // Sort for consistent output
        files.sort();
        
        Ok(files)
    }

    pub fn get_branches(&self) -> Result<Vec<String>> {
        let repo = self.repo.as_ref().ok_or(GitScannerError::Git("No repository open".to_string()))?;
        
        let mut branches: Vec<_> = repo.branches(None, git2::BranchTarget::Local)?
            .filter_map(|b| b.ok())
            .map(|b| b.name().unwrap_or("").to_string())
            .collect();
        
        // Sort and filter out HEAD references
        branches.sort();
        branches.retain(|s| !s.starts_with("HEAD"));
        
        Ok(branches)
    }

    pub fn get_tags(&self) -> Result<Vec<String>> {
        let repo = self.repo.as_ref().ok_or(GitScannerError::Git("No repository open".to_string()))?;
        
        let mut tags: Vec<_> = repo.tags()?
            .filter_map(|t| t.ok())
            .map(|t| t.name().unwrap_or("").to_string())
            .collect();
        
        // Sort and filter out annotated vs lightweight duplicates
        tags.sort();
        tags.dedup();
        
        Ok(tags)
    }

    pub fn generate_report(&self, report: &ScanReport) -> Result<String> {
        let mut output = String::new();
        
        output.push_str("=== Pipewatch Pro Git Scanner Report ===\n\n");
        
        // Summary section
        output.push_str(&format!("Summary:\n"));
        output.push_str(&format!("  Repository: {:?}\n", report.repo_url.as_deref().unwrap_or("Unknown")));
        output.push_str(&format!("  Total Commits: {}\n", report.commit_count));
        output.push_str(&format!("  Unique Authors: {}\n", report.total_authors));
        
        if let Some(last_date) = report.last_commit_date {
            output.push_str(&format!("  Last Commit: {}\n", last_date.format("%Y-%m-%d %H:%M:%S UTC")));
        } else {
            output.push_str("  Last Commit: Unknown\n");
        }
        
        // OWASP Findings
        if !report.owasp_findings.is_empty() {
            output.push_str("\n--- OWASP Top 10 Findings ---\n");
            
            let mut by_severity: HashMap<SeverityLevel, Vec<_>> = HashMap::new();
            for finding in &report.owasp_findings {
                by_severity.entry(finding.severity).or_insert_with(Vec::new).push(finding);
            }
            
            for (severity, findings) in &by_severity {
                let severity_str: String = (*severity).into();
                output.push_str(&format!("\n  [{}] {}\n", severity_str, findings.len()));
                
                for finding in findings.iter().take(5) {
                    output.push_str(&format!("    - {} (Commit: {:?})\n", 
                        &finding.description[..100],
                        finding.commit_id.as_deref(),
                    ));
                }
            }
        } else {
            output.push_str("\n--- OWASP Top 10 Findings ---\n");
            output.push_str("  No OWASP patterns detected\n");
        }
        
        // Suspicious Patterns
        if !report.suspicious_patterns.is_empty() {
            output.push_str(&format!("\n--- Suspicious Patterns ({}) ---\n", report.suspicious_patterns.len()));
            
            let mut by_type: HashMap<String, &SuspiciousPattern> = HashMap::new();
            for pattern in &report.suspicious_patterns {
                by_type.entry(pattern.pattern_type.clone()).or_insert(pattern);
            }
            
            for (type_name, pattern) in &by_type {
                output.push_str(&format!("  [{}] - {}\n", type_name, &pattern.description));